django/db/models/functions/text.py

```diff
@@ -188,6 +191,121 @@ def as_sqlite(self, compiler, connection, **extra_context):
         return super().as_sql(compiler, connection, function='UNICODE', **extra_context)
 
 
+class RegexpFlagMixin:
+    default_flags = {
+        'mariadb': 'c',  # Force case-sensitive matching.
+        'mysql': 'c',  # Force case-sensitive matching.
+        'postgresql': 'p',  # Force partial newline-sensitive matching.
+    }
+    inline_flags = {}
+    wrap_pattern = {}
+
+    @cached_property
+    def flag_mapping(self):
+        return {k: str.maketrans(v) for k, v in {
+            'mariadb': {'c': '-i'},
+            'mysql': {'s': 'n'},
+            'oracle': {'s': 'n'},
+            'postgresql': {'m': 'n'},
+        }.items()}
+
+    def as_sql(self, compiler, connection, **extra_context):
+        expression, pattern, *extra, flags = self.get_source_expressions()
```

Comment:
You can avoid all of the expression creation below by creating a copy of `self` before hand.

```suggestion
        copy = self.copy()
        expression, pattern, *extra, flags = copy.get_source_expressions()
```

And then `copy.set_source_expressions([expression, pattern, *extra])` and `return super(RegexpFlagMixin, copy).as_sql(compiler, connection, **extra_context)`.

Reply:
Thanks Simon.

Do I understand you correctly that the other changes above are unnecessary if I make this one?

Reply:
yep, as long as don't alter `self` you should be good; you an alter a local copy as much as you'd like.